### ISSUE TITLE
ikhal takes a long time to start (1.5 seconds)

### ISSUE THREAD

**Issue by edi9999** [pimutils/khal#564]
I would expect a command line programm to start in less than 200ms

When I run the following command : 

```
strace -tt ikhal 2>/tmp/straced.txt
```

I get in straced.ikhal

[straced.txt](https://github.com/pimutils/khal/files/752204/straced.txt)

**Comment by hobarrera** [comment]
Running khal in a clean virtualenv (one with khal and its dependencies) takes 0.294s. Running it via the system python, it take 1.003s.

I've seen similar results from other python apps (including `todoman`), but haven't really figured out what the problem it (wild guess: do too many python packages slow things down?).

**Comment by geier** [comment]
agree, this isn't good at all. Given some of the libraries we are currently using, we will not be able to bring it down by a factor of 10 whatever we do. Even thoug, in your example, a really high amount of time seems to go into importing various python libraries. Could you try:

    IMPORTS="pytz urwid click tzlocal"
    for one in $IMPORTS; do echo $one;  -c "import $one"; done

I get:

    pytz
    
    real    0m0.254s
    user    0m0.238s
    sys     0m0.017s
    urwid
    
    real    0m0.086s
    user    0m0.086s
    sys     0m0.000s
    click
    
    real    0m0.060s
    user    0m0.052s
    sys     0m0.008s
    tzlocal
    
    real    0m0.160s
    user    0m0.120s
    sys     0m0.040s

We can probably get rid of the time for tzlocal by only importing it if really needed.

**Comment by edi9999** [comment]
$ for one in $IMPORTS; do echo $one;  time python -c "import $one"; done
pytz

real    0m0.062s
user    0m0.048s
sys     0m0.014s
urwid

real    0m0.125s
user    0m0.105s
sys     0m0.019s
click

real    0m0.090s
user    0m0.072s
sys     0m0.017s
tzlocal

real    0m0.060s
user    0m0.044s
sys     0m0.016s


All times seem to be very small.

**Comment by hobarrera** [comment]
Indeed, the largest of these times is less than 5%.

I tried [asking](http://stackoverflow.com/q/42050450/107510) in SO why the system python might be so much slower than a virtualenv one, but haven't had luck. I'm honestly out of ideas here.

**Comment by untitaker** [comment]
Main slowdown is usually `pkg_resources`, adding a overhead of 200-300ms to todoman and khal. I work around this by avoiding the setuptools entry points, which import that module.

Try running `python -m cProfile -o result $(which khal) --help` and browsing the profile data with snakeviz. Ignoring `pkg_resources` there's a lot of stuff in there that really doesn't need to be imported when outputting the help page.

**Comment by untitaker** [comment]
Oh, one thing: It helps if you do `pip install .` instead of `pip install -e .`. This avoids usage of `pkg_resources`.

**Comment by hobarrera** [comment]
Both for `khal` and `todoman`, the setuptools entry point (eg: `pkg_resources`) adds a 60-80% overhead for me (it varies in that range on my machines).

I used this script to compare:

```
$ cat k.py 
from khal.cli import main_khal

main_khal()
```

Unless it's just me seeing the results, I'd look for an alternative to this.

> Oh, one thing: It helps if you do `pip install .` instead of `pip install -e .`. This avoids usage of `pkg_resources`.

This doesn't work nicely for downstream packages.

**Comment by hobarrera** [comment]
[Upstream issue](pypa/setuptools#510)

**Comment by edi9999** [comment]
The correct link is https://github.com/pypa/setuptools/issues/510

**Comment by untitaker** [comment]
>This doesn't work nicely for downstream packages.

Could you elaborate? As far as I know, distro packages shouldn't use `-e` anyway and therefore not depend on `pkg_resources`.

**Comment by hobarrera** [comment]
Downstream doesn't use `-e`, but *does* depend on `pkg_resources`.

My previous comment meant to point out that `pip install -e` can't be used by downstream packagers, so that idea won't help.

**Comment by untitaker** [comment]
How does downstream depend on pkg_resources? At least Arch uses setuptools (like pip without `-e`) to generate the executables.

**Comment by hobarrera** [comment]
> How does downstream depend on pkg_resources? At least Arch uses setuptools (like pip without -e) to generate the executables.

The generated script (created via `python setup.py install`) uses `pkg_resources`:

```
$ cat $(which khal)
#!/usr/bin/python
# EASY-INSTALL-ENTRY-SCRIPT: 'khal==0.9.2.dev14+ngcb5c6c4.d20170205','console_scripts','khal'
__requires__ = 'khal==0.9.2.dev14+ngcb5c6c4.d20170205'
import re
import sys
from pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
        load_entry_point('khal==0.9.2.dev14+ngcb5c6c4.d20170205', 'console_scripts', 'khal')()
    )
```
